File: Hash/Hash02.cpp
#include <string>
#include <vector>
#include <set>
#include <iostream>

using namespace std;

bool solution(vector<string> phone_book) {
    bool answer = true;
    string left,right;
    multiset<string> st(phone_book.begin(),phone_book.end());
    for(auto it=st.begin();it!=st.end();it++){
        for(auto it2=it;it2!=st.end();it2++){           
            if(it!=it2){                                        // 같은 위치의 값을 방지하여 비교
                left=*it;
                right=*it2;
                if(right.find(left)==0){                        // string의 find함수는 위치를 반환시키므로 접두사 즉 첫 번째 위치일 경우 0을 반환
                    return false;}
            }
        }
    }
    return answer;
}

File: Hash/Hash02.java
import java.util.Arrays;
import java.util.HashMap;

class Solution {
    public boolean solution(String[] phone_book) {
        boolean answer = true;

        Arrays.sort(phone_book);
        HashMap<String, Integer> hashMap = new HashMap<>();

        for(String i: phone_book) {
            hashMap.put(i, 0);
        }

        for(int former = 0; former < phone_book.length - 1; former++) {
            for(int latter = former + 1; latter < phone_book.length; latter++) {
                String smaller = phone_book[former];
                if(phone_book[latter].contains(smaller)) {
                    hashMap.put(phone_book[latter], hashMap.get(phone_book[latter]) + 1);
                }
            }
        }

        for(String phoneNumber: phone_book) {
            if(hashMap.get(phoneNumber) > 0) {
                answer = false;
            }
        }
        return answer;
    }
}